File: src/main/java/com/freya02/botcommands/api/pagination/interactive/BasicInteractiveMenu.java
package com.freya02.botcommands.api.pagination.interactive;

import com.freya02.botcommands.api.components.InteractionConstraints;
import com.freya02.botcommands.api.components.event.StringSelectionEvent;
import com.freya02.botcommands.api.pagination.TimeoutInfo;
import com.freya02.botcommands.api.pagination.paginator.BasicPaginator;
import com.freya02.botcommands.api.utils.ButtonContent;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.interactions.components.selections.SelectOption;
import net.dv8tion.jda.api.interactions.components.selections.StringSelectMenu;
import net.dv8tion.jda.api.utils.messages.MessageEditData;
import net.dv8tion.jda.internal.utils.Checks;
import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * @param <T> Type of the implementor
 */
@SuppressWarnings("unchecked")
public abstract class BasicInteractiveMenu<T extends BasicInteractiveMenu<T>> extends BasicPaginator<T> {
	protected final List<InteractiveMenuItem<T>> items;

	protected int selectedItem = 0;
	protected final boolean usePaginator;

	protected BasicInteractiveMenu(InteractionConstraints constraints, TimeoutInfo<T> timeout, boolean hasDeleteButton,
	                               ButtonContent firstContent, ButtonContent previousContent, ButtonContent nextContent, ButtonContent lastContent, ButtonContent deleteContent,
	                               @NotNull List<InteractiveMenuItem<T>> items, boolean usePaginator) {
		super(constraints, timeout, 0, (a, b, c, d) -> new EmbedBuilder().build(), hasDeleteButton, firstContent, previousContent, nextContent, lastContent, deleteContent);

		if (items.isEmpty()) throw new IllegalStateException("No interactive menu items has been added");

		this.usePaginator = usePaginator;
		this.items = items;
		setSelectedItem(0);
	}

	@NotNull
	protected StringSelectMenu.Builder createSelectMenuBuilder() {
		final LambdaStringSelectionMenuBuilder builder = componentss.stringSelectionMenu(this::handleSelection).oneUse().setConstraints(constraints);

		final List<SelectOption> options = builder.getOptions();
		for (int i = 0, itemsSize = items.size(); i < itemsSize; i++) {
			InteractiveMenuItem<T> item = items.get(i);

			SelectOption option = item.content().toSelectOption(String.valueOf(i));
			if (i == selectedItem) option = option.withDefault(true);

			options.add(option);
		}

		return builder;
	}

	private void handleSelection(StringSelectionEvent event) {
		selectedItem = Integer.parseInt(event.getValues().get(0));

		event.editMessage(get()).queue();
	}

	public int getSelectedItem() {
		return selectedItem;
	}

	public SelectContent getSelectedItemContent() {
		return items.get(selectedItem).content();
	}

	/**
	 * Sets the interactive menu item number, <b>this does not update the embed in any way</b>,
	 * you can use {@link #get()} with an <code>editOriginal</code> in order to update the embed on Discord
	 *
	 * @param itemIndex Index of the item, from <code>0</code> to <code>[the number of menus] - 1</code>
	 *
	 * @return This instance for chaining convenience
	 */
	public T setSelectedItem(int itemIndex) {
		Checks.check(itemIndex >= 0, "Item index cannot be negative");
		Checks.check(itemIndex < items.size(), "Item index cannot be higher than max items count (%d)", items.size());

		this.selectedItem = itemIndex;
		setMaxPages(items.get(itemIndex).maxPages());
		setPage(0);

		return (T) this;
	}

	/**
	 * Sets the interactive menu item number, via it's label (O(n) search), <b>this does not update the embed in any way</b>,
	 * you can use {@link #get()} with an <code>editOriginal</code> in order to update the embed on Discord
	 *
	 * @param itemLabel Label of the item, must be a valid label from any of the interactive menu items
	 *
	 * @return This instance for chaining convenience
	 */
	public T setSelectedItem(String itemLabel) {
		Checks.notEmpty(itemLabel, "Item name cannot be empty");

		for (int i = 0; i < items.size(); i++) {
			final String label = items.get(i).content().label();

			if (label.equals(itemLabel)) {
				return setSelectedItem(i);
			}
		}

		throw new IllegalArgumentException("Item name '" + itemLabel + "' cannot be found in this interactive menu");
	}

	@Override
	public MessageEditData get() {
		onPreGet();

		if (usePaginator) {
			putComponents();
		}

		components.addComponents(createSelectMenuBuilder().build());

		final MessageEmbed embed = items.get(selectedItem).supplier().get((T) this, getPage(), messageBuilder, components);
		messageBuilder.setEmbeds(embed);
		messageBuilder.setComponents(components.getActionRows());

		onPostGet();

		return messageBuilder.build();
	}
}

File: src/main/java/io/github/freya022/botcommands/api/pagination/interactive/BasicInteractiveMenu.kt
package io.github.freya022.botcommands.api.pagination.interactive

import io.github.freya022.botcommands.api.components.Components
import io.github.freya022.botcommands.api.components.data.InteractionConstraints
import io.github.freya022.botcommands.api.components.event.StringSelectEvent
import io.github.freya022.botcommands.api.pagination.PaginatorSupplier
import io.github.freya022.botcommands.api.pagination.TimeoutInfo
import io.github.freya022.botcommands.api.pagination.paginator.BasicPaginator
import io.github.freya022.botcommands.api.utils.ButtonContent
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.interactions.InteractionHook
import net.dv8tion.jda.api.interactions.components.selections.SelectOption
import net.dv8tion.jda.api.interactions.components.selections.StringSelectMenu
import net.dv8tion.jda.api.utils.messages.MessageEditData
import net.dv8tion.jda.internal.utils.Checks

/**
 * @param T Type of the implementor
 */
abstract class BasicInteractiveMenu<T : BasicInteractiveMenu<T>> protected constructor(
    componentsService: Components,
    constraints: InteractionConstraints,
    timeout: TimeoutInfo<T>?,
    hasDeleteButton: Boolean,
    firstContent: ButtonContent,
    previousContent: ButtonContent,
    nextContent: ButtonContent,
    lastContent: ButtonContent,
    deleteContent: ButtonContent,
    @JvmField
    protected val items: List<InteractiveMenuItem<T>>,
    @JvmField
    protected val usePaginator: Boolean
) : BasicPaginator<T>(
    componentsService,
    constraints,
    timeout,
    0,
    PaginatorSupplier { _, _, _, _ -> EmbedBuilder().build() },
    hasDeleteButton,
    firstContent,
    previousContent,
    nextContent,
    lastContent,
    deleteContent
) {
    var selectedItem: Int = 0
        protected set
    val selectedItemContent: SelectContent
        get() = items[selectedItem].content

    init {
        check(items.isNotEmpty()) { "No interactive menu items has been added" }

        setSelectedItem(0)
    }

    protected fun createSelectMenu(): StringSelectMenu {
        val options = arrayOfNulls<SelectOption>(items.size)
        var i = 0
        val itemsSize = items.size
        while (i < itemsSize) {
            val item = items[i]

            var option = item.content.toSelectOption(i.toString())
            if (i == selectedItem) option = option.withDefault(true)

            options[i] = option
            i++
        }

        return componentsService.ephemeralStringSelectMenu()
            .bindTo { event: StringSelectEvent -> this.onItemSelected(event) }
            .oneUse(true)
            .constraints(constraints)
            .addOptions(*options)
            .build()
    }

    private fun onItemSelected(event: StringSelectEvent) {
        selectedItem = event.values[0].toInt()

        event.editMessage(get()).queue()
    }

    /**
     * Sets the interactive menu item number, **this does not update the embed in any way**,
     * you can use [get] with an [InteractionHook.editOriginal] in order to update the embed on Discord
     *
     * @param itemIndex Index of the item, from `0` to `[the number of menus] - 1`
     *
     * @return This instance for chaining convenience
     */
    fun setSelectedItem(itemIndex: Int): T {
        Checks.check(itemIndex >= 0, "Item index cannot be negative")
        Checks.check(itemIndex < items.size, "Item index cannot be higher than max items count (%d)", items.size)

        this.selectedItem = itemIndex
        maxPages = items[itemIndex].maxPages
        page = 0

        return this as T
    }

    /**
     * Sets the interactive menu item number, via it's label (O(n) search), **this does not update the embed in any way**,
     * you can use [get] with an [InteractionHook.editOriginal] in order to update the embed on Discord
     *
     * @param itemLabel Label of the item, must be a valid label from any of the interactive menu items
     *
     * @return This instance for chaining convenience
     */
    fun setSelectedItem(itemLabel: String): T {
        Checks.notEmpty(itemLabel, "Item name cannot be empty")

        for (i in items.indices) {
            val label = items[i].content.label

            if (label == itemLabel) {
                return setSelectedItem(i)
            }
        }

        throw IllegalArgumentException("Item name '$itemLabel' cannot be found in this interactive menu")
    }

    override fun get(): MessageEditData {
        onPreGet()

        if (usePaginator) {
            putComponents()
        }

        components.addComponents(createSelectMenu())

        val embed = items[selectedItem].supplier.get(this as T, page, messageBuilder, components)
        messageBuilder.setEmbeds(embed)
        messageBuilder.setComponents(components.actionRows)

        onPostGet()

        return messageBuilder.build()
    }
}
